src/tools/cephfs_mirror/SnapReplicator.cc

```diff
@@ -0,0 +1,434 @@
+        return -1;
+      }
+    } else {
+      std::clog << __func__ << ":" << __LINE__ << ":  failed to readlink(" << dentry << ")" << std::endl;
+      my_errno = -rv;
+      return -1;
+    }
+  } else {
+    std::clog << __func__ << ":" << __LINE__ << ":  unhandled entry " << de.d_name << std::endl;
+  }
+  return 0;
+}
+
+int SnapReplicator::replicate()
+{
+  bool is_error = false;
+  const struct timespec epoch_rctime{0,0};
+  const struct timespec old_rctime = get_rctime(m_old_snap, m_src_dir + "/"s + m_src_snap_dir + "/"s + m_old_snap);
+  std::string new_snap_root = m_src_dir + "/"s + m_src_snap_dir + "/"s + m_new_snap;
+  struct ceph_statx old_stx;
+  struct ceph_statx new_stx;
+  struct ceph_dir_result *new_dirp = NULL;
+  struct dirent de;
+  bool is_first_sync = (old_rctime == epoch_rctime);
+  std::queue<std::string> dir_queue;
+
+  std::clog << __func__ << ":" << __LINE__ << ": old rctime:"
+    << old_rctime.tv_sec << "."
+    << std::setw(9) << std::setfill('0')
+    << old_rctime.tv_nsec
+    << std::endl;
+
+  if (ceph_chdir(m_src_mnt, new_snap_root.c_str()) < 0) {
+    std::clog << __func__ << ":" << __LINE__ << ": unable to chdir to " << new_snap_root << std::endl;
+    return -1;
+  } else {
+    std::clog << __func__ << ":" << __LINE__ << ": chdir to " << new_snap_root << std::endl;
+  }
+
+  dir_queue.push(""s);
+  // replicate the changes
+  // we traverse the new snap only in dirs with (mtime|ctime) > old snap ctime
+  while (!is_error && !dir_queue.empty()) {
+    std::string dir_name = dir_queue.front();
+    dir_queue.pop();
+
+    if (ceph_opendir(m_src_mnt, ("."s + dir_name).c_str(), &new_dirp) < 0) {
+      std::clog << __func__ << ":" << __LINE__ << ": opendir failed for " << ("." + dir_name) << std::endl;
+      is_error = true;
+      break;
+    }
+    ceph_assert(std::string(ceph_readdir(m_src_mnt, new_dirp)->d_name) == ".");
+    ceph_assert(std::string(ceph_readdir(m_src_mnt, new_dirp)->d_name) == "..");
+
+    while (ceph_readdirplus_r(m_src_mnt, new_dirp, &de, &new_stx, CEPH_STATX_BASIC_STATS, AT_SYMLINK_NOFOLLOW, NULL) > 0) {
+      if (!is_first_sync) {
+        std::string old_dentry = m_src_dir      + "/"s +
+                                 m_src_snap_dir + "/"s +
+                                 m_old_snap     +
+                                 (dir_name == "" ? ""s : (dir_name + "/"s)) +
+                                 (dir_name == "" ? "/"s + de.d_name : de.d_name);
+
+        int rv = 0;
+        if ((rv = ceph_statx(m_src_mnt, old_dentry.c_str(), &old_stx, CEPH_STATX_BASIC_STATS, AT_SYMLINK_NOFOLLOW)) == 0) {
+          // found entry in old snap
+          std::clog << __func__ << ":" << __LINE__ << ": processing old entry " << old_dentry << std::endl;
+          if (handle_old_entry(dir_name, old_dentry, de, old_stx, new_stx, old_rctime, dir_queue) < 0) {
+            is_error = true;
+            break;
+          }
+        } else if (rv == -ENOENT) {
+          goto new_entry;
```

Comment:
try to avoid using `goto` if possible. I think with a bit of reorganization it won't be needed here.

Reply:
fixed